### ISSUE TITLE
route-href binding too late

### ISSUE THREAD

**Issue by EisenbergEffect** [aurelia/templating-router#46]
@fkleuver commented on [Mon Nov 28 2016](https://github.com/aurelia/router/issues/438)

**I'm submitting a bug report**

* **Library Version:**
1.0.7

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
6.9.1

* **NPM Version:**
4.0.2

* **Aurelia-CLI**
0.22.0

* **Browser:**
all

* **Language:**
TypeScript 2.0.10

**Current behavior:**
Given an application with a recursive route `'app'` (but it seems to apply to nested child routes in general) where navigation occurs through `a` tags with `route-href` attributes, the following happens:
- Click on the first route and it navigates to `'#/app'`
- Click on the second route and it navigates to `'#/app/app'`
- Click on the third route and it navigates to `'#/app/app/app'`

Then after a refresh:
- Click on the third route and it navigates to `'#/app'`

**Expected/desired behavior:**
After the refresh:
- Click on the third route and it navigates to `'#/app/app/app'`

**repro**
This is from a question in gitter.
The person who asked about this (@cabarney) pinpointed it down to a wrong order of sequence in which the `route-href` is databound when refreshing:
1. _refreshBaseUrl on the first level route
2. _refreshBaseUrl on the second level route
3. generate on the router for the third level route
4. _refreshBaseUrl on the third level route

So I tried to force step 3. to come after step 4. by moving `this.processChange();` from `bind()` to `attached()` in `route-href.js` and that seems to solve this particular problem.

I've made a minimal repro here:
https://github.com/fkleuver/aurelia-route-href-issue

You can see the difference by clicking `App` several times, refreshing, then clicking the last `App` again. The normal `route-href` will fail, but the `route-href-attached` will work.

I don't know if moving `this.processChange()` to `attached` will break anything else though.

**EDIT**
Sorry, this issue was meant to go to `templating-router`. Is it possible to move it?

**Comment by EisenbergEffect** [comment]
@fkleuver I don't think there are any really issues from moving it to attached. I would be interested to find out why this is necessary. I'm wondering if there's something wrong elsewhere that is causing this. Are you interested in investigating further? If not, would you at least be willing to submit a PR for the timing change in the route-href attribute?

**Comment by fkleuver** [comment]
@EisenbergEffect Interesting.. I think I found the problem:

`CommitChangesStep` comes after `ActivateNextStep` in the router pipeline. `CommitChangesStep`
refreshes the baseUrl on the router, among other things. This baseUrl is needed for `route-href` to get the correct relative url.

Here's where it goes wrong: `ActivateNextStep` kicks off the ViewModel lifecycle without actually waiting for the rest of the router pipeline to complete. So views are already being created and binded while `CommitChangesStep` is still doing stuff. This means there is no guarantee that baseUrl (and potentially other properties) are complete/correct by the time the ViewModel lifecycle starts.

This is generally not a problem because with one-by-one navigation the router is always "one ahead" of the viewmodel. However, when directly navigating to the full url the nested routes are all loaded at once and then this race condition pops up. It happens consistently for all child routes which are nested more than 2 levels deep. 

Furthermore, simply wrapping `this.processChange()` in a `queueMicroTask` instead of placing it inside `attached()` also solves the problem. This makes sense (to me at least) because it does enough to postpone `this.processChange()` until the router pipeline is fully done.

That being said, I don't think it's `route-href` that needs to be fixed but rather something in the router pipeline. But that might bring additional ramifications with it.

What's your opinion on this?

**Comment by EisenbergEffect** [comment]
Good analysis. Thanks for taking the time to look into that. In the worst case, we can delay the href set until attached. That's not a huge deal. Ideally, we want to delay route generation until right after the base url is fixed. Maybe we can introduce a hook for this. We might just use attached in the short term and then switch to another mechanism in the future. I need to consider it more...

**Comment by fkleuver** [comment]
Hm, that brought me on a different idea. I tried using the BindingEngine to subscribe to changes to `router.baseUrl` and call `processChanges()` when that happens. It works like a charm. Besides, wouldn't you want to update the hrefs when the router's baseUrl changes anyway?

The only two downsides I can think of are the added dependency on `aurelia-binding` and potentially an impact on performance. I don't know if that matters much though. If it doesn't, I can create a PR for that. Otherwise I'll stick to moving it to `attached`. Let me know..

**Comment by paulharker** [comment]
Hi all,
I have a heavily child-router based application that is suffering from this bug. We are in the framework phase, but within a few weeks to a month will start moving to a feature-creation phase. Would love to know how tis fix is coming. Paul

**Comment by bryanrsmith** [comment]
@fkleuver @EisenbergEffect Do you guys know if this issue is blocked on a decision? I agree with @fkleuver's assessment that the router's commit step happening at the end of the lifecycle is the underlying issue. Since that's fairly involved to change at this point it would be great to work around the issue in route-href to at least get some people unblocked. What do you guys think of simply delaying the generation as Rob suggests, pending further discussions? I can help out with this if needed (but it sounds like there may already be a couple alternatives implemented already?)

**Comment by EisenbergEffect** [comment]
I think we can move this forward by delaying until after attached for now. @bryanrsmith If you have some time to make the fix, that would be awesome.

**Comment by matthewh** [comment]
What is needed to get this merged? I am using @bryanrsmith's fix and it works well.

**Comment by shahidkarimi** [comment]
Please answer this
https://stackoverflow.com/questions/47441179/aurelia-href-bind-is-not-working
